### DISCUSSION TITLE
bv_maker failed

### DISCUSSION THREAD

**Discussion by Mhaalk** [brainvisa/brainvisa.github.io#43] [Q&A] [answered]
Hello,

Je suis sur Ubuntu 18.04, j'essaye d'installer le _developer environnement_ avec des modules privés (le baby pipeline de f. leroy).

j'ai installé le paquet singularity 3.7 distribué sur brainvisa.info
j'ai réussi à lancer singularity sans problème.

une fois arrivé à l'étape de "bv_maker configure" je reçois l'erreur suivante:

```
Summary of Git repositories in /casa/host/src
=============================================
┌───── * uncommitted working tree changes
│┌──── + uncommitted index changes
││┌─── $ stash present
│││┌── % untracked files
││││ ┌ HEAD
││││ │                 ┌ update status
││││ │                 │ ┌ bv_maker upstream state
││││ │                 │ │  ┌ Git upstream state
││││ │                 │ │  │┌ Directory
     master (88948f70) ✓ U=  aims/aims-free/master
     master (d6e740c)  ✓ U=  aims/aims-gpl/master
     master (fc376ac)  ✓ U=  aims/aims-til/master
     master (bdc5ce40) ✓ U=  anatomist/anatomist-free/master
     master (f467771)  ✓ U=  anatomist/anatomist-gpl/master
     master (d5d55dfd) ✓ U=  axon/master
     master (8a30b54)  ✓ U=  brainvisa-share/master
     master (76676b5)  ✓ U=  capsul/master
     master (966d223)  ✓ U=  development/brainvisa-cmake/master
     master (f85abb4)  ✓ U=  development/brainvisa-release/master
     master (de2a8fd)  ✓ U=  development/casa-distro/master
     master (15899a1)  ✓ U=  highres-cortex/master
     master (750f4ad)  ✓ U=  morphologist/morpho-deepsulci/master
     master (c1cdb6c)  ✓ U=  morphologist/morphologist-ui/master
     master (b92ab3f)  ✓ U=  populse/populse_db/master
     master (cbb2ad7)  ✓ U=  soma/soma-base/master
     master (4b36785)  ✓ U=  soma/soma-io/master
     master (26532b4)  ✓ U=  soma/soma-workflow/master
Traceback (most recent call last):
  File "/casa/host/bootstrap/brainvisa-cmake/bin/bv_maker", line 1104, in process
    **meth_kwargs)
  File "/casa/host/bootstrap/brainvisa-cmake/bin/bv_maker", line 3736, in configure
    self.process_configuration_lines()
  File "/casa/host/bootstrap/brainvisa-cmake/bin/bv_maker", line 3494, in process_configuration_lines
    'components_sources.json')) as f:
IOError: [Errno 2] No such file or directory: u'/brainvisa/source/components_sources.json'
Skipping build of /casa/host/build because it depends on a step that failed.

bv_maker summary:
OK          step sources: /casa/host/src, started: 2021/04/29 15:01, stopped: 2021/04/29 15:01
FAILED      step configure: /casa/host/build, started: 2021/04/29 15:01, stopped: 2021/04/29 15:01
UNMET DEP   step build: /casa/host/build
There were errors.
```

j'ai essayé grossièrement de contourner le problème en changeant le chemin qui se trouve physiquement à `$HOME/casa_distro/brainvisa-opensource-master/src/components_sources.json`

mais ça n'a généré que plus d'erreurs... comment puis-je contourner ce problème?

**Comment by sapetnioc** [comment]
This is weird. Could you please copy the commands you used to setup the environment and launch `bv_maker`.

**Reply by Mhaalk** [reply]
I did:

```
#to install Golang before singularity:
tar -C /usr/local -xzf go1.16.3.linux-amd64.tar.gz 
echo 'export PATH=$PATH:/usr/local/go/bin' >> ~/.bashrc

#to install dependencies required by singularity:
sudo apt-get update && sudo apt-get install -y \
    build-essential \
    uuid-dev \
    libgpgme-dev \
    squashfs-tools \
    libseccomp-dev \
    wget \
    pkg-config \
    git \
    cryptsetup-bin

#then installed this:
https://brainvisa.info/download/singularity-container-3.7.0-ubuntu-18.04.deb


#then:
mkdir -p $HOME/casa_distro/brainvisa-opensource-master

singularity run -B $HOME/casa_distro/brainvisa-opensource-master:/casa/setup $HOME/casa_distro/casa-dev-5.0-1.sif branch=master distro=opensource

echo 'export PATH="$HOME/casa_distro/brainvisa-opensource-master/bin:$PATH""' >> ~/.bashrc

bv bash
bv_maker
```

**Reply by denisri** [reply]
You don't need to install go and its dependencies before singularity: the `.deb` is ready to install; go is used to compile singularity (and build the `.deb`, on our side), not to use it.

**Comment by denisri** [comment]
Hi,
We reply in english because other people outside of France may read/use the questions and answers and avoid asking again the same...
The path `/brainvisa/source/...` seems very weird, does not exist, and I don't know how it came here, I don't think it is mentioned anywhere in the code of casa-distro... Did you edit the config file `/casa/host/conf/bv_maker.cfg` manually and add this path anywhere ?

**Reply by Mhaalk** [reply]
I did edit bv_maker.cfg manually like this, using the cfg from a previous build:

might it be because of line 3 from the build section?
```

[ source $CASA_SRC ]
  brainvisa opensource bug_fix
  + perso/fleroy perso/fleroy/bug_fix
  + brainvisa/cortical_surface/cortical_surface-private/branches/bug_fix cortical_surface/cortical_surface-private/bug_fix
  + brainvisa/morphologist/morphologist-private/branches/bug_fix morphologist/morphologist-private/bug_fix
  + baby bug_fix
  - optical_imaging
  - nuclear_imaging
  - fMRI-*
  - web

[ build $CASA_BUILD ]
  build_type = bug_fix
  make_options = -j8
  opensource bug_fix /brainvisa/source
  + /brainvisa/source/cortical_surface/cortical_surface-private/bug_fix
  + /brainvisa/source/morphologist/morphologist-private/bug_fix
  + /brainvisa/source/morphologist/baby/bug_fix  
  + /brainvisa/source/perso/fleroy/bug_fix
  - connectomist-*   
  - brainrat
  - optical_imaging
  - nuclear_imaging
  - snapbase
  - fMRI-*
  - axon_web
```

**Comment by denisri** [comment, marked as answer]
> [ build $CASA_BUILD ]
>   build_type = bug_fix
>   make_options = -j8
>   opensource bug_fix /brainvisa/source

OK that's it: the paths you wrote in the config file don't exist in the container. You should use `$CASA_SRC` instead of `/brainvisa/source`.

**Reply by Mhaalk** [reply]
oh thanks my bad! 
Now I corrected like this:

I did not understand what -j8 ask Make to do and so how to correct my command (as I were again from my previous .cfg)
```
[ source $CASA_SRC ]
  brainvisa opensource bug_fix
  + perso/fleroy perso/fleroy/bug_fix
  + brainvisa/cortical_surface/cortical_surface-private/branches/bug_fix cortical_surface/cortical_surface-private/bug_fix
  + brainvisa/morphologist/morphologist-private/branches/bug_fix morphologist/morphologist-private/bug_fix
  + baby bug_fix
  - optical_imaging
  - nuclear_imaging
  - fMRI-*
  - web

[ build $CASA_BUILD ]
  build_type = bug_fix
  make_options = -j8
  opensource bug_fix $CASA_SRC
  + $CASA_SRC/cortical_surface/cortical_surface-private/bug_fix
  + $CASA_SRC/morphologist/morphologist-private/bug_fix
  + $CASA_SRC/morphologist/baby/bug_fix  
  + $CASA_SRC/perso/fleroy/bug_fix
  - connectomist-*   
  - brainrat
  - optical_imaging
  - nuclear_imaging
  - snapbase
  - fMRI-*
  - axon_web
```
but then bv_maker throw me this:

```
Building directory: /casa/host/build
$ make -j8
make: *** No targets specified and no makefile found.  Stop.
Traceback (most recent call last):
  File "/casa/host/bootstrap/brainvisa-cmake/bin/bv_maker", line 1104, in process
    **meth_kwargs)
  File "/casa/host/bootstrap/brainvisa-cmake/bin/bv_maker", line 4023, in build
    timeout=timeout)
  File "/casa/host/bootstrap/brainvisa-cmake/bin/bv_maker", line 428, in system
    six.reraise(*error)
  File "/casa/host/bootstrap/brainvisa-cmake/bin/bv_maker", line 417, in system
    raise OSError()
OSError: Command failed: u'make' '-j8'
From directory: /casa/host/build

bv_maker summary:
FAILED      step build: /casa/host/build, started: 2021/04/29 16:40, stopped: 2021/04/29 16:40
There were errors.
```

**Comment by denisri** [comment]
`-j8` tells make to run up to 8 compile jobs in parallel because you seem to have 8 CPU cores on your machine. It should be OK.
Here you ran `bv_maker build`, but did the `configure` step run without error before that (after you fixed your configuration) ? If this hasn't been done, no wonder the build does not work ;)

**Reply by Mhaalk** [reply]
Indeed I forgot to relaunch the configure first... now I did and build went all good, thanks!

now, but still I receive warnings at Brainvisa start now:
would that impact any function ?

```
maxime@dell:~$ bv
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-maxime'
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-maxime'
Loading toolbox baby
Loading toolbox cortical_surface
Loading toolbox data management
Loading toolbox datamind
Loading toolbox freesurfer
Loading toolbox morphologist
Loading toolbox spm
Loading toolbox structural_analysis
Loading toolbox tools
Loading toolbox viewers
Loading toolbox My processes
soma-workflow starting in light mode
Workflow controller initialised
The log file for this session is '/casa/home/.brainvisa/brainvisa.log'

Database /casa/host/build/share/brainvisa-share-5.1 must be updated because
there is no database file.

/usr/local/lib/python2.7/dist-packages/matplotlib/__init__.py:910: MatplotlibDeprecationWarning: The backend.qt5 rcParam was deprecated in version 2.2.  In order to force the use of a specific Qt binding, either import that binding first, or set the QT_API environment variable.
  mplDeprecation)


ERROR: **

    
    
    ImportError: /casa/host/build/brainvisa/toolboxes/cortical_surface/processes/anatomy/tools/ParcelsTextureFromCoordinates.py: No module named sigraph

**

* * *

`**ImportError**`

    
    
    Traceback (most recent call last):
      File "/casa/host/build/python/brainvisa/processes.py", line 6363, in addDir
        toolbox=toolbox)  # two arguments : process fullpath and category (directories separated by /)
      File "/casa/host/build/python/brainvisa/processes.py", line 5948, in readProcess
        sys.exc_info()[2])
      File "/casa/host/build/python/brainvisa/processes.py", line 5933, in readProcess
        moduleName, fileIn, fileName, moduleDescription)
      File "/casa/host/build/brainvisa/toolboxes/cortical_surface/processes/anatomy/tools/ParcelsTextureFromCoordinates.py", line 32, in <module>
        import sigraph
    ImportError: /casa/host/build/brainvisa/toolboxes/cortical_surface/processes/anatomy/tools/ParcelsTextureFromCoordinates.py: No module named sigraph
    



ERROR: **
```

**Comment by denisri** [comment]
Considering your `bv_maker.cfg`, you have added some non-opensource projects, but not all of them and there are missing dependencies. You'd rather consider using the "`bvdev`" distro or project group instead of "`opensource`", or add at least the `sulci_private` component.

**Reply by Mhaalk** [reply]
Hello,
thanks to your guidance I can run BV from GUI perfectly now, but I still receive this:

`(base) maxime@dell:~$ bv
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-maxime'
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-maxime'
Loading toolbox baby
Loading toolbox cortical_surface
Loading toolbox data management
Loading toolbox datamind
Loading toolbox freesurfer
Loading toolbox morphologist
Loading toolbox spm
Loading toolbox structural_analysis
Loading toolbox tools
Loading toolbox viewers
Loading toolbox My processes
soma-workflow starting in light mode
Workflow controller initialised
The log file for this session is '/casa/home/.brainvisa/brainvisa.log'

/usr/local/lib/python2.7/dist-packages/matplotlib/__init__.py:910: MatplotlibDeprecationWarning: The backend.qt5 rcParam was deprecated in version 2.2.  In order to force the use of a specific Qt binding, either import that binding first, or set the QT_API environment variable.
  mplDeprecation)
exec: /casa/host/build/brainvisa/toolboxes/morphologist/startup.py
connect to computing resource: dell
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
`

my guess is these are minor error that won't really bother me, do you agree or should I correct it all?

------------------------------------------------------------------------------------------------------------------------------------

also, (but maybe this shoud be a distinct thread)

I'm now trying to run python scripts that require to import anatomist and soma, I sourced bv_env.sh (both by an export in .bashrc, and by a source ... command, but IPython throw me:

`---> 12 from soma import aims

~/casa_distro/brainvisa-opensource-master/build/python/soma/aims/__init__.py in <module>
    131
    132 try:
--> 133     from soma.aims import aimssip
    134 except ImportError as exc:
    135     if exc.args[0] == ('dynamic module does not define init function '

ImportError: libpython2.7.so.1.0: cannot open shared object file: No such file or directory
`

**Comment by denisri** [comment]
For brainvisa, yes these are minor warnings that are not harmful, it's OK.

For the error while loading soma.aims, I suspect you are trying to load the module from the host system, without getting into the container (which seems to indicate the path "`~/casa_distro/brainvisa-opensource-master/build/python/soma/aims/init.py`" which is a host path, not a container path). The programs are built inside a container, using a python installed inside the container, thus everything has to run inside the container also. Normally i's just a matter of running `bv ipython` instead of `ipython`. `brainvisa` does it automatically, because it's not installed elsewhere on the system, so we provide a run scripts which does the equivalent of `bv brainvisa` under the hood (same for all commands `anatomist`, `AimsSomething` etc. But we can't do that for `python`, which exists and is needed on the host system: we can't overload it otherwise we would break the host system.

Denis
